### PR TITLE
Mac specific details

### PR DESCRIPTION
**Pull request by ChrisWhealy**
Hi Aaron et al

Apologies for the large-scale rewrite, but I figured that since I had to spend some time pulling things apart to understand how these examples worked, others would probably have to do the same - so the MD files for the first two examples are simply an explanation of what's going on, based on the results of my somewhat deeper analysis.

Also, I "prettied-up" the code samples to make them more "beginner friendly"

If anything needs to be changed, I'm happy to adjust it.

Chris W

### CODE REVIEW

**Comment by torch2424** [comment]
@ChrisWhealy 

Woah! 😄 Thank you soooooo much for taking the time to fill in the gaps in the docs! This is super appreciated! 

Reviewing this all now 😄

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 1]
```diff
@@ -6,69 +6,201 @@ sidebar_label: Transforming WASI Modules
...
+
+1. Start `parcel` 
+
+   ```bash
+   $ parcel index.html
+   ```
+
+1. Point your browser to [`http://localhost:1234`](http://localhost:1234) and you should see `Standard Output: Done!` appear both on the browser screen and in the JavaScript console
+
+# The `clock_time_get` WebAssembly Module
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: So personally I don't think this should live in the docs here. We wanted to focus on running WebAssemlbly modules in these docs, rather than explaining them, and teaching WebAssembly. Teaching WebAssembly, and how to build/write it itself we were planning on doing somewhat later in its own section.

Also, since we are examining the `.wat` here, which is not really meant for an end user to have to read/write, I feel like this can be placed in a separate tutorial or something later down the line 😄 👍 

cc @syrusakbary for the judgement call here.

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 2]
```diff
@@ -6,69 +6,201 @@ sidebar_label: Transforming WASI Modules
...
+
+Two things are important to notice here:
+
+1. The `import` keyword indicates that function `clock_time_get` lives in an external module called `wasi_unstable`
+
+1. The interface to this function is described by the previously declared type definition `$t0`.  In other words, we know for certain that function `clock_time_get` must be passed an `i64` as its second parameter.
+
+Having established exactly what the interface is to `clock_time_get`, we now know that we cannot call this WASM module without first transforming it.
+
+### Important
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: So I went ahead an opened an issue at: https://github.com/wasmerio/docs.wasmer.io/issues/24

That way the `clock_time_get` module makes a bit more sense for an end-user 😄

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 3]
```diff
@@ -6,69 +6,201 @@ sidebar_label: Transforming WASI Modules
...
+```JavaScript
+const loweredWasmBytes = await lowerI64Imports(wasmBytes)
+```
+
+Here, the `lowerI64Imports` function transforms the interface such that a JavaScript `BigInt` value can be transferred to a WebAssembly `i64` value as an array of 8, unsigned, 8-bit integers.
+
+It is not until after this transformation has occurred that we can instantiate the WebAssembly module and invoke it as before.
+
+```JavaScript
+// *****************************************************************************
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: Same comment here about the dividers 👍

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 4]
```diff
@@ -6,69 +6,201 @@ sidebar_label: Transforming WASI Modules
...
+
+The coding seen below is very similar to the coding used for the previous `hello-world` example with the following important difference.
+
+Inside function `startWasiTask`, we fetch the WASM file contents and convert it to a `Uint8Array` as before, but then there is the additional line
+
+```JavaScript
+const loweredWasmBytes = await lowerI64Imports(wasmBytes)
+```
+
+Here, the `lowerI64Imports` function transforms the interface such that a JavaScript `BigInt` value can be transferred to a WebAssembly `i64` value as an array of 8, unsigned, 8-bit integers.
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: This is a great TL;DR explanation of the code! Thank you! 😄

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md [hunk 5]
```diff
@@ -6,113 +6,166 @@ sidebar_label: Hello World
...
 
-In this example, we will run the WASI module [as-echo](https://github.com/torch2424/as-echo), using `@wasmer/wasi`. 
 
-The WASI module will echo "Hello World", and we will get the standard output from `/dev/stdout` using the `@wasmer/wasmfs` module. 
+In this example, we want to use the following call chain:
 
-This example will be run in the browser, bundled and served by [Parcel](https://parceljs.org/). 
+`JavaScript` --> `WebAssembly` --> `Some native "OS" function`
 
-However, `@wasmer/wasi` and `@wasmer/wasmfs` also work in Node, and the code examples from this guide can be used as a loose example, as long as the ES6 syntax is replaced with what would be the node equivalent.
```
[the remark this answers is not in the document]
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Yeah, its rather vague.  I'll remove this comment and re-explain this fundamental point.
My plan is to create equivalent examples for using wasmer-js from Node - but this will be a separate PR...

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md [hunk 6]
```diff
@@ -6,113 +6,166 @@ sidebar_label: Hello World
...
-npm init
+$ npm install -g parcel
 ```
 
-Then, answer the prompts from `npm init`. After this, we will install the `parcel-bundler` package, as well as `parcel-plugin-static-files-copy` to allow serving our wasm files as a static asset:
+> ### Mac users
+> Before the installation of Parcel will work on a Mac, you must first install the [Xcode Command Line Tools](https://developer.apple.com/download/more/?=for%20Xcode)
 
-`npm install --save-dev parcel-bundler parcel-plugin-static-files-copy`
+## Run Example from Git Clone
```
[the remark this answers is not in the document]
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Well OK, but if any assumptions being made, they should be clearly stated.

I'll reinstate the step-by-step instructions rather than recommending to clone the repo...

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md [hunk 7]
```diff
@@ -6,113 +6,166 @@ sidebar_label: Hello World
...
-const wasmFilePath = './as-echo.wasm';
+## JavaScript Coding
+
+Seeing as this is demo code, it uses meaningful variable names and contains lots of explanatory comments (features that are often sadly missing from production code).  Please take some time to read and understand these comments as they explain how the functionality has been constructed.
+    
+Also, make a note of the comment explaining `@wasmer/wasm-transformer`; we will cover this very important detail in a later example.
+
+
+```JavaScript
+// *****************************************************************************
```
[the remark this answers is not in the document]
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Well, to avoid going round in circles on this subject, when creating code specifically for the purpose of education, I've found that it helps students if there is some sort of visual separator between the different blocks of code.

How visually intrusive these separators are can be controlled by the choice of character:

* A line of asterisks stands out the strongest
    ```// *******************************```

* A line of equals signs is slightly less intrusive
    ```// ===============================```

* A line of minus signs is less intrusive
    ```// -------------------------------```

* A line of alternating minus signs and spaces is the least intrusive
    ```// - - - - - - - - - - - - - - - -```

If it doesn't bug you too much, I'd like to leave these separators in, but downgrade them to the least visually intrusive...  🙏🏻

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 8]
```diff
@@ -6,69 +6,201 @@ sidebar_label: Transforming WASI Modules
...
+In the previous Hello World example, we showed you how to run the very basic `as-echo` WASM module that received a text string as an argument and simply echoed it back via standard out.  However, some WASI modules may be compiled in a way that means they can't immediately be run from a JavaScript environment such as a browser.
 
-Thus, in order to run these modules, you would have to **transform the module** with `@wasmer/wasm-transformer`.
+For example, any module that calls the [clock\_time\_get](https://github.com/NuxiNL/cloudabi/blob/master/cloudabi.txt#L1230) WASI API, must be able to supply a 64-bit, signed integer.  However, passing a JavaScript `BigInt` to a WebAssembly `i64` is not yet supported &mdash; this detail is still at the proposal stage.  (See [here](https://github.com/WebAssembly/JS-BigInt-integration/issues/15) and [here](https://github.com/WebAssembly/proposals/issues/7) for details).
```
[the remark this answers is not in the document]
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: Updated...

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 9]
```diff
@@ -6,69 +6,201 @@ sidebar_label: Transforming WASI Modules
...
 
-In this example, we will fetch a WASI module, transform it with `lowerI64Imports` from `@wasmer/wasm-transformer`, and then run it in the browser!
+However, it is not impossible to run such a module; but before we can, we must first ***transform*** it using `@wasmer/wasm-transformer`.
 
-```javascript
+> ### Under The Hood  
+> Technically, this transformation adapts the interface so that JavaScript `BigInt`s (64-bit, signed integers) in can be transferred to and from WebAssembly in the form of arrays of 8, unsigned 8-bit integers (a JavaScript `Uint8Array`)
+
+## Prerequisites
```
[the remark this answers is not in the document]
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: Rearranged the instructions to repeat the steps shown in "hello world", but with certain modifications

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md [hunk 10]
```diff
@@ -6,113 +6,166 @@ sidebar_label: Hello World
...
 
-This example will be run in the browser, bundled and served by [Parcel](https://parceljs.org/). 
+`JavaScript` --> `WebAssembly` --> `Some native "OS" function`
 
-However, `@wasmer/wasi` and `@wasmer/wasmfs` also work in Node, and the code examples from this guide can be used as a loose example, as long as the ES6 syntax is replaced with what would be the node equivalent.
+`WASI` is the bridge that allows a WebAssembly module to invoke a native "OS" function.
 
-First, let's set up a small JavaScript project using Parcel. To do this, we will create a new directory, and initialize it as an node project with npm:
+> ### As an Aside...  
+> The term "OS" is in double quotes to indicate that the native function being called might not actually belong to the underlying operating system.
```
[the remark this answers is not in the document]
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Since this explanation applies generically to all languages, I've move it to the top level introduction.md

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (torch2424) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: @ChrisWhealy Thank you! 😄

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (torch2424) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: @ChrisWhealy Ah, well if you've found them to be helpful in education in the past, we can totally keep them 😄

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md (hunk 1 above)
Comment (torch2424) [reply docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: @ChrisWhealy Thank you for the thoughtful  explanation 😄 You make some very good points, I'm 50 / 50 on this now. Thank you for moving it, but, if anything, we can move it back if @syrusakbary is also convinced 👍

File: docs/wasmer-js/cli/user-guide.md [hunk 11]
```diff
@@ -4,12 +4,12 @@ title: Wasmer-JS Command Line Interface User Guide
 sidebar_label: User Guide
 ---
 
-```
+```bash
```
Comment (torch2424) [inline docs/wasmer-js/cli/user-guide.md]: Thank you for adding the bash specifiers! 😄 🎉

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md [hunk 12]
```diff
@@ -1,118 +1,199 @@
...
 sidebar_label: Hello World
 ---
 
 [Full Example Project Source Code](https://github.com/wasmerio/docs.wasmer.io/tree/master/docs/wasmer-js/node-modules/examples/hello-world)
 
-In this example, we will run the WASI module [as-echo](https://github.com/torch2424/as-echo), using `@wasmer/wasi`. 
+# Hello World!
 
-The WASI module will echo "Hello World", and we will get the standard output from `/dev/stdout` using the `@wasmer/wasmfs` module. 
+In this very simple example, we want to develop a Browser-based application that calls a WebAssembly module that in turn calls a native "OS" function.
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: So I try to avoid using words like "simple", as it can be discouraging for developers when something is "simple", but they are having a hard time with it 😄

Not a blocker, but something we can go back in another PR and fix 👍

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md [hunk 13]
```diff
@@ -1,118 +1,199 @@
...
-- `@wasmer/wasm` - which provides a sandboxed filesystem that our wasi module will interact with
+1. Change into some development directory
 
-To install these packages as dependencies to the project, run the following:
+    ```bash
+    $ cd <some_development_directory>
+    ```
 
-`npm install --save @wasmer/wasi @wasmer/wasmfs`
+1. Create and then change into a new project directory, then run `npm init`
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Thank you for explaining the step by step again! 😄

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/how-do-i-know-if-a-module-needs-transformation.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
+---
+id: wasmer-js-modules-how-do-i-know-if-a-module-needs-transformation
+title: Wasmer-JS in the Browser
+sidebar_label: How Do I Know if a Module Needs Transformation?
+---
+
+# How Do I Know if a WASM Module Needs Transformation?
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/transforming-wasi-modules/how-do-i-know-if-a-module-needs-transformation.md]: This is perfect! 😄 Thank you sooooo much! 😄

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md [hunk 15]
```diff
@@ -1,74 +1,125 @@
...
 sidebar_label: Transforming WASI Modules
 ---
 
 [Full Example Project Source Code](https://github.com/wasmerio/docs.wasmer.io/tree/master/docs/wasmer-js/node-modules/examples/transforming-wasi-modules)
 
-In the Hello World example, we covered how to run a basic "echo" wasm module. 
+# Why Is Transformation Necessary?
 
-However, some WASI modules may be compiled in a way that can't be yet run by the browser. For example, modules that call the [clock_time_get](https://github.com/WebAssembly/WASI/blob/master/phases/old/snapshot_0/docs/wasi_unstable.md#clock_time_get) WASI API, require passing a `BigInt` from Javascript as a `I64` into WebAssembly, which is currently not yet fully supported and still in a [proposal for the browser](https://github.com/WebAssembly/JS-BigInt-integration/issues/15). 
+In the previous Hello World example, we showed you how to run the very basic `as-echo` WASM module that received a text string as an argument and simply echoed it back via standard out.  However, some WASI modules may be compiled in a way that means they can't immediately be run from a JavaScript environment such as a browser.
```
Comment (torch2424) [inline docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: Same here with words like "basic", but again, can fix in a followup PR 😄 👍

**Comment by torch2424** [comment]
Wanted to see how it looked on my machine real quick, and looks great! 😄 👍 

Screen Shot 2020-01-22 at 12 03 45 PM

Merging for real now 😄

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md (hunk 1 above)
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: Sure.

I'm not too bothered ***where*** this explanation goes (put it in multiple places if you want).  I do contend the following though:
* The explanation must be provided
* The explanation should be put in a place (or places) where it would naturally arise in someone's train of thought

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md (hunk 12 above)
Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Hmmm....

If a concept is described as "simple", yet people struggle to understand it, then I would contend that the explanation is at fault because it has failed to transmit the underlying simplicity - as Albert Einstein said *"If you can't explain it to a six year old, then you don't understand it yourself"*

Further to this, I would also contend that there exists the widespread misunderstanding that the words "simple" and "easy" are synonyms.  This is simply not true and is the basis for a lot of people giving up after only a few steps of their learning journey.  "Simple" means minimum complexity, "Easy" means minimum effort - the two are ***not*** the same (See Rich Hickey's excellent explanation of this https://www.youtube.com/watch?v=oytL881p-nQ)

That said, I'm happy to swap the word "simple" for "basic" and then ensure that the explanations do not require the reader to make any leaps of understanding.

**Comment by ChrisWhealy** [comment]
I tested all the changes in Docasaurus myself…

:-)

Chris Whealy

M +44 (0)7808 575377 
Twitter: @LogaRhythm

"The voice of ignorance speaks loud and long,
  but the words of the wise are quiet and few"
                                                 Ancient Proverb

File: docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md (hunk 1 above)
Comment (torch2424) [reply docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: Ah okay, if you are not worried where it goes, I definitely agree with the two points. I think where it currently is is good, and if people open issues, or have trouble running through the docs, lets move it / link to it 😄 👍

Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/transforming-wasi-modules/transforming-wasi-modules.md]: 👍🏼

File: docs/wasmer-js/node-modules/examples/hello-world/hello-world.md (hunk 12 above)
Comment (torch2424) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Ah, so I hope you don't mind these discussions. Just for a quick check in, totally am a super friendly person, and again am so happy to have your contributions, and be able to discuss this 😄 

So some quick background on me: I don't have a ton of experience in the education field, but I've been really passionate about writing documentation for the past few years, so I've been reaching out for all the resources I can on how / what makes good documentation good, and bad documentation bad. So our discussions have been immensely helpful for me, thank you 😄 And like not just saying that to be nice, or like because I work for the company, but like for real personally saying that 👍 

That being said, I do think you make some very good points:

> If a concept is described as "simple", yet people struggle to understand it, then I would contend that the explanation is at fault because it has failed to transmit the underlying simplicity - as Albert Einstein said "If you can't explain it to a six year old, then you don't understand it yourself"

Using the word "simple" on a complex explanation definitely is the fault of the explanation. Let's say this is Point 1.

> Further to this, I would also contend that there exists the widespread misunderstanding that the words "simple" and "easy" are synonyms. This is simply not true and is the basis for a lot of people giving up after only a few steps of their learning journey. "Simple" means minimum complexity, "Easy" means minimum effort - the two are not the same (See Rich Hickey's excellent explanation of this https://www.youtube.com/watch?v=oytL881p-nQ)

I also agree here, "Simple" and "Easy" do not mean the same thing. Let's consider this Point 2.

> That said, I'm happy to swap the word "simple" for "basic" and then ensure that the explanations do not require the reader to make any leaps of understanding.

So I think I have a suggestion along the same line, but without using another "similar" word 😄 

So from my experience of learning how to write good docs, a friend of mine, Carolyn Stransky, gave a really good talk on "Humanizing your documentation". In the talk, she mentioned [using the words "simple" or "easy"](https://youtu.be/FvPXMUuUCS4?t=1244) is bad practice. And I've seen this echoed in multiple talks in the JS, and general development, community. This is because, I try to write documentation from the audience's point of view, and our audience can be: people of different skill levels, people who are already frustrated looking for additional help, and people who may not be able to understand english. And like you mentioned in point 2, even though it is wrong, people think simple and easy are the same thing, so it can be discouraging or frustrating if your a junior developer who doesn't understand, a senior developer who spent the past 2 hours debugging and is now trying to understand why it doesn't work, or someone trying to "google translate" the docs to their native language.

So as Carolyn stated in their talk, maybe instead of just avoiding the word "simple", maybe we can say something like: "In this example, we will to develop a Browser-based application that calls a WebAssembly module. This in turn calls a native "OS" function, to output the string "Hello World"".

What do you think? 😄

Comment (ChrisWhealy) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: Totally with you on this...

I've spent the last 25 years writing stuff like software training courses, technical documentation, trade-show presentations (even written two books) and in all this, I am continually striving to meet the three criteria of:

* Simplicity  (as per the above definition of "simple")
* Clarity
* Brevity

So I totally agree with your motivation and *modus operandi*, albeit, we have some differences in terminology.

Here's a suggestion, seeing as its 8:30PM here in the YooKay, could you merge the changes for https://github.com/wasmerio/docs.wasmer.io/pull/25, then on Monday, I'll have a shiny new PR waiting for you that removes words like "basic" etc.  The reason for a new PR is that I've seen a few more annoying typos and clumsy phrasing that also need to be smoothed out...

What say ye?

Comment (torch2424) [reply docs/wasmer-js/node-modules/examples/hello-world/hello-world.md]: @ChrisWhealy That sounds perfect! Thank you! 😄 I am actually testing out #25 now haha! Will merge it in like 10 minutes 😄
